### DISCUSSION TITLE
JAX for key functions for autodiff and GPU acceleration

### DISCUSSION THREAD

**Discussion by shishirdholakia** [acoular/acoular#553] [Ideas]
I'd love to know if you've been thinking about adding an optional [JAX](https://github.com/google/jax) backend for specific Acoular functions to enable autodiff and GPU acceleration.

The current NumPy/SciPy/numba framework limits performance and prevents gradient-based optimization.

Rewriting specific functions using JAX equivalents (nearly identical API) for NumPy/SciPy would:

1. Speed up beamforming and CSM computations via GPU execution and JIT compilation
2. Allow differentiable workflows (e.g., optimizing mic positions, steering weights, etc)
3. Provide compatibility with ML frameworks like Flax for NN and numpyro for bayesian inference

As a start, JAX equivalents of the following could be implemented:

- steering_vectors()
- beamform_power()
- csm computation

I don't have any background in beamforming and audio, but it looks like many of these could be achieved with simple drop-in replacements of jax.numpy and jax.scipy. One drawback is to use the jax functions, you would likely have to load the arrays into memory. This would probably be doable on most modern compute platforms and reasonable mic array sizes, but you can probably chunk arrays if it isn't. 

Would love to know if there are any plans for this or if you think it would be useful!

**Comment by esarradj** [comment]
At the moment most of the heavy lifting inside Acoular is done using carefully optimized JIT compiled functions. JIT is based on [numba](https://numba.readthedocs.io/en/stable/#). It does not provide autodiff, though.

To better understand if JAX could be an alternative: are you aware of any comparison between our present approach and JAX? This would be very helpful. When we decided to use numba in the first place we did some research and back then we found out that numba outperformed the alternatives (e.g. Cython) in our case.

**Comment by artpelling** [comment]
> I'd love to know if you've been thinking about adding an optional [JAX](https://github.com/google/jax) backend for specific Acoular functions to enable autodiff and GPU acceleration.

It is something that we have considered a couple of years ago. So far, we did not fully investigate this option.

> The current NumPy/SciPy/numba framework limits performance and prevents gradient-based optimization.

What do you mean by this? Certainly, gradient-based optimization is possible for most problems.

> Rewriting specific functions using JAX equivalents (nearly identical API) for NumPy/SciPy would:
> 
> 1. Speed up beamforming and CSM computations via GPU execution and JIT compilation
> 2. Allow differentiable workflows (e.g., optimizing mic positions, steering weights, etc)
> 3. Provide compatibility with ML frameworks like Flax for NN and numpyro for bayesian inference
> 
> As a start, JAX equivalents of the following could be implemented:
> 
> * steering_vectors()
> * beamform_power()
> * csm computation

Note that we currently use Numba for JIT-compilation to accelerate the computations that you mentioned (steering vectors, CSM). There is a possibility of enabling Cuda with Numba https://github.com/acoular/acoular/issues/140, so GPU acceleration could also be achieved this way.

> * beamform_power()

What do you mean here?

> I don't have any background in beamforming and audio, but it looks like many of these could be achieved with simple drop-in replacements of jax.numpy and jax.scipy.

Firstly, let me say that Acoular is not limited to acoustics. The only thing that makes it specific to acoustics is the [standard value for the wave speed](https://www.acoular.org/api_ref/generated/acoular.environments.html#acoular.environments.Environment.c) which can be changed easily. The drop-in syntax of `jax` is very nice. However, a change here would probably be more involved if you want the entire computation chain to be accelerated by `jax` because we use Numba-accelerated functions in many places.

> One drawback is to use the jax functions, you would likely have to load the arrays into memory.

I am not sure what you mean here. Maybe you could elaborate. Note that Acoular uses a block-wise pipeline processing approach which means that the time data is processed in chunks (which live in the memory).

> This would probably be doable on most modern compute platforms and reasonable mic array sizes, but you can probably chunk arrays if it isn't.

So far, we did not have an application where the array size became an issue. The computational bottleneck is usually with "simple" DSP-computations on the raw data. Roughly, about how many sensors in an array are we talking about with the application that you have in mind?

> Would love to know if there are any plans for this or if you think it would be useful!

As said, we do not have any concrete plans, yet, but are very open to new directions and improvements to the package. Nonetheless, adding dependencies to a package is always something that increases maintenance effort. So there needs to be an obvious and significant benefit of doing so. As it stands I do not see this immediately (we would need to compare the speed of `jax` not to pure Python but to our Numba-accelerated functions which have already been optimized quite a lot over the course of the last years).
 
We are always interested in applications that Acoular is or can be used for (especially in fields other than aero-acoustics). Describing your application and the problems/bottlenecks that you face with the current version in more detail would be very valuable. Generally, ML applications and Bayesian inference methods are interesting to us. @adku1173 works in this field. Maybe you also want to have a look at https://github.com/adku1173/acoupipe if you are interested in generating training data for ML.
